### ISSUE TITLE
VCI Expert Review and Approval - Provisional Interpretation

### ISSUE THREAD

**Issue by jimmyzhen** [ClinGen/clincoded#1423]
Implement the **Provisional Interpretation** section, which consists of the following requirements:
1. Section panel title: **Provisional Interpretation**
2. A dropdown of **by Individual** or **on behalf of Group** options for the user to select.
3. Dropdown label: **Interpretation being saved 'by Individual' or 'on behalf of Group'**
4. Display of form fields for labels **ClinGen Group Name** _OR_ **Non-ClinGen Group Name**. Either one of these fields is required if the **on behalf of Group** option is selected.
5. A **Save as Provisional Interpretation** button for the user to save the selected option from the dropdown.
6. The dropdown selection is **required** if the user wants to save the provisional interpretation.
7. Display of a note with message: **Note: Saving as a Provisional Interpretation does not submit it to ClinVar.**
8. Display of the **Current Saved Provisional Interpretation**, including the timestamp.

**Comment by wrightmw** [comment]
@jimmyzhen ClinVar VariationID: 10000 does not load data in this instance. It doesn't load in this test instance but it does work in the normal demo and prod interfaces.

@jimmyzhen ClinVar VariationID: 17000 did work in the test instance when I created it as part of 10001 Affiliation but when I went back in and tried to enter VariationID: 17000 as part of Affiliation 10002 it would no longer load any data.
In other words, it seems like this loading issue arises when a curator from an Affiliation tries to enter a ClinVar VariationID that has already been entered by a curator from a different Affiliation.

**Comment by selinad** [comment]
@wrightmw @jimmyzhen I'm thinking we don't need this text any more (on Evaluation Summary)

![image](https://user-images.githubusercontent.com/11320314/38337075-de1b3b86-3818-11e8-817f-ad06890c8591.png)

**Comment by selinad** [comment]
@jimmyzhen @wrightmw I may be confused -- I thought it was possible to save an Interoperation as LP or P, just not as Provisional.  I got this trying to save going around the 2nd time after an Approved

![image](https://user-images.githubusercontent.com/11320314/38337581-bfddedd8-381a-11e8-9c52-ffaa83b0c994.png)

**Comment by jimmyzhen** [comment]
> I thought it was possible to save an Interoperation as LP or P, just not as Provisional. I got this trying to save going around the 2nd time after an Approved

@selinad, in regards to your comments above, you need to provide an explanation to the modified pathogenicity before you can save the interpretation.

**Comment by selinad** [comment]
@jimmyzhen If I get a warning I can't save as Provisional then go back and add a disease and View Summary again, it gives me the Provisional box.  The provisional box should only appear after a Save of the Classification

![image](https://user-images.githubusercontent.com/11320314/38338630-2818da94-381f-11e8-99af-2abca9fce6a8.png)

**Comment by selinad** [comment]
Need a space between sentences in this text:

![image](https://user-images.githubusercontent.com/11320314/38339425-f27df906-3822-11e8-97b1-391fb17607fd.png)

**Comment by selinad** [comment]
@jimmyzhen If I had Modified to LP, then go back around and add more and get Calculated of P, then it still says "LP" for Modification, but if I change to P it makes me give a reason...is the same logic being used in GCI and VCI? thx!

![image](https://user-images.githubusercontent.com/11320314/38340038-244f6f2a-3826-11e8-8207-227276dd7dc7.png)

**Comment by jimmyzhen** [comment]
@selinad, I don't know whether they are using the very same logic, although they kinda behave similarly.

**Comment by selinad** [comment]
@jimmyzhen here's what it does in GCI

![image](https://user-images.githubusercontent.com/11320314/38340390-132af7d0-3828-11e8-9197-01350f96f4cd.png)

**Comment by wrightmw** [comment]
@jimmyzhen An empty "Provisional Classification entered by" field in editable box is filled upon saving. Perhaps get rid of this from editable box and only display once saved?

**Comment by wrightmw** [comment]
-[] @jimmyzhen I got confirmation back from the curators... they want "date reviewed" for the editable date field for Provisional Interpretation and "date approved" for the editable date field for Approved Interpretation.

**Comment by wrightmw** [comment]
- [] @jimmyzhen The yellow text needs to be updated from "...thisInterpretation..." to "...this Interpretation..."
![screen shot 2018-04-05 at 3 23 28 pm](https://user-images.githubusercontent.com/15131169/38395012-ab4e0030-38e5-11e8-85cd-10ed5da8b49b.png)

**Comment by wrightmw** [comment]
- [ @] @jimmyzhen The "Provisional Classification entered by:" should only appear after submitting:

![screen shot 2018-04-05 at 3 21 24 pm](https://user-images.githubusercontent.com/15131169/38395071-f40e0f0e-38e5-11e8-8c11-0c6f932f0da4.png)

**Comment by wrightmw** [comment]
- [ ] @jimmyzhen This yellow text also needs a gap:

![screen shot 2018-04-05 at 3 29 19 pm](https://user-images.githubusercontent.com/15131169/38395129-4c10403c-38e6-11e8-9287-c04f07b122f5.png)

**Comment by wrightmw** [comment]
It's looking pretty fantastic so far @jimmyzhen ... you've done an amazing job with this!

**Comment by selinad** [comment]
- [ ] @jimmyzhen I Approved an Interpretation (looks fantastic!) and then added more evidence and came back to test as it was now Pathogenic.  Saved and it did the right thing to ask me to add Disease.  Did that came back and changed to In Progress -- maybe it happened when I saved w/o a disease (will check)

![image](https://user-images.githubusercontent.com/11320314/38396067-e703b44e-38ea-11e8-9d19-1880ce628ba4.png)

**Comment by selinad** [comment]
@jimmyzhen adding a 2nd Prov and 2nd Approved Interpretation works GREAT!
![image](https://user-images.githubusercontent.com/11320314/38396118-28d20f9c-38eb-11e8-8351-19cd535e0345.png)

**Comment by selinad** [comment]
- [ ] @jimmyzhen I can't load the same variant as a different curator than myself still -- not sure whether this bug was addressed?

**Comment by wrightmw** [comment]
@selinad I don't think that bug is fixed because I am getting the same issue

**Comment by wrightmw** [comment]
@selinad re. text above... 

Suggested text:
"When ready, you may save this Provisional Interpretation as Approved. Once you have saved it as Approved it will become uneditable, but you will be able to save a new current Approved Interpretation, thus archiving this current one and retaining access to its Evaluation Summary."

**Comment by selinad** [comment]
Looks good @wrightmw -- thank you!

**Comment by selinad** [comment]
- [ ] @jimmyzhen for saved Provisional and Approved snapshot boxes, change "Classification" to "interpretation" (VCI Summary page)

- [ ] @jimmyzhen when Provisional exists and not Approved and someone comes back to Summary, the Evaluation Summary box stays open like it is, but Approval box needs to disappear and button goes on snapshot for most current Saved Provisional (per GCI workflow)

- [ ] documents show new Status buttons for header

**Comment by selinad** [comment]
- [ ] @jimmyzhen everything is looking great!  All the things we talked about look to be working well on first run-through.  Only one potential problem is after delete disease -- then something that was approved goes back to in progress:

![image](https://user-images.githubusercontent.com/11320314/38464209-56b494aa-3abe-11e8-9e6e-e4590cf4a6c9.png)

**Comment by selinad** [comment]
@jimmyzhen I think the above also occurs if I add a disease back....

![image](https://user-images.githubusercontent.com/11320314/38464243-fab92e26-3abe-11e8-815c-f069f02c7013.png)

**Comment by selinad** [comment]
p.s. love the "Approve this Saved Provisional" button!!!!

**Comment by wrightmw** [comment]
And I am now able to work on the same variant as part of different Affiliations. Great!

**Comment by wrightmw** [comment]
@jimmyzhen @selinad 

When you first visit an Interpretation, you get to see "All Existing Interpretations", which is great, but this is not behaving quite right.

Here are the headers for the existing Interpretations I randomly created:

1) An APPROVED "likely pathogenic" interpretation from the Hereditary Cancer WG Affiliation
![screen shot 2018-04-08 at 6 46 57 pm](https://user-images.githubusercontent.com/15131169/38475612-2d5f71b0-3b5e-11e8-9acb-4416d58c758d.png)

2) An APPROVED/NEW PROVISIONAL "benign" interpretation by the Rett/Angelman 
![screen shot 2018-04-08 at 6 48 04 pm](https://user-images.githubusercontent.com/15131169/38475625-40718ae0-3b5e-11e8-8beb-89a871cf701d.png)

3) A PROVISIONAL VUS interpretation by the RASopathy EP
![screen shot 2018-04-08 at 6 51 44 pm](https://user-images.githubusercontent.com/15131169/38475668-813e24de-3b5e-11e8-9af1-92d0b54a6283.png)

4) An IN PROGRESS interpretaion by the KCNQ1 EP.. for this one I have only Saved an Evaluation Summary... nothing else
![screen shot 2018-04-08 at 7 00 58 pm](https://user-images.githubusercontent.com/15131169/38475813-51b1a276-3b5f-11e8-83d0-7e0bd26d5760.png)

5) I started an Interpretation as BCM-Education...added evidence ...but I have not saved an Evaluation Summary yet... so no status
![screen shot 2018-04-08 at 7 31 12 pm](https://user-images.githubusercontent.com/15131169/38476608-85986c60-3b63-11e8-83cc-a06807dcc057.png)

Now when I go in as a different Affiliation, this time as "No Affiliation" and I go to the Evidence View for the above 5 examples, here is my header:
![screen shot 2018-04-08 at 7 35 32 pm](https://user-images.githubusercontent.com/15131169/38476755-4e4bd746-3b64-11e8-9c59-f08af1eee243.png)

- [ ] Rett/Angelman says "PROVISIONAL" but it is actually "APPROVED/NEW PROVISIONAL"

- [ ] For BCM-Education because I have not saved an Evaluation yet, it is not marked as IN PROGRESS in the headers but when I go back to my dashboard (acting as BCM-Education) it says "In progress":

![screen shot 2018-04-08 at 7 32 40 pm](https://user-images.githubusercontent.com/15131169/38476661-d697f6c6-3b63-11e8-8536-718404ba627d.png)

Therefore, we are inconsistently showing the "In Progress" status.

**Comment by wrightmw** [comment]
Btw, as you can see I can definitely work on the same variant as different Affiliations. Whatever you did with the elastic search is working! Great!

**Comment by selinad** [comment]
@jimmyzhen @wrightmw this is what I saw before.  I started an Interpretation, saved as BA1 Met -- you can see it responds I saved, but I don't get update button.  Also criteria bar does not change...
@jimmyzhen let me know if I should make a ticket....I believe a curator was writing in about this and I asked for more info.

![image](https://user-images.githubusercontent.com/11320314/38510538-384e386e-3bda-11e8-8c35-3cb8b102e9e5.png)

I then try to do PM2 and have the same problem:
![image](https://user-images.githubusercontent.com/11320314/38510900-39d3e93a-3bdb-11e8-83a1-a205cbd7a394.png)

**Comment by selinad** [comment]
@jimmyzhen when I clicked on the orange Approved button, it went back to In progress and I was not able to get the chance to Approve:

![image](https://user-images.githubusercontent.com/11320314/38513893-76091692-3be4-11e8-9da3-98386053ee09.png)

**Comment by selinad** [comment]
@jimmyzhen @wrightmw  If I go go back to Summary to Approve a saved Provisional that was benign after adding more evidence and it's now pathogenic, I get this, which is very nice.
![image](https://user-images.githubusercontent.com/11320314/38530282-8b92cbe0-3c1e-11e8-8baf-6e0532c428fa.png)


However, if I click to save it as Approved (orange button), I get this, which seems correct, but we may need to make certain curators aren't thinking they are saving the pathogenic....I'm not sure what to do for this, or whether we should worry right now?

![image](https://user-images.githubusercontent.com/11320314/38530340-c1713094-3c1e-11e8-90ee-433fa512eedb.png)

**Comment by selinad** [comment]
@jimmyzhen steps for buggy thing
- saved a Provisional as benign
- modified Interpretation so it was Pathogenic and needed a disease
- added a diesease
- returned to summary
- clicked orange button to Approve previous Provisional
- got this: (looks like it is not bringing up Approval box, but a Provisional one....

![image](https://user-images.githubusercontent.com/11320314/38530521-da7a13de-3c1f-11e8-9636-f02f3367eb65.png)

**Comment by selinad** [comment]
@jimmyzhen thoughts: 

-we have this text in GCI -- analogous text could be helpful in eliminating confusion in VCI:
![image](https://user-images.githubusercontent.com/11320314/38535042-88d70c96-3c36-11e8-85e9-02bcae19358d.png)

- the white is hard to see in the orange bar...should we make this text black so we can see Provisional vs. Approved?

**Comment by jimmyzhen** [comment]
@selinad,

> the white is hard to see in the orange bar...should we make this text black so we can see Provisional vs. Approved?

I made the text bold and it seems to read better now:
![image](https://user-images.githubusercontent.com/6956310/38541084-cffd8caa-3c52-11e8-9dde-67e14c93a123.png)

> we have this text in GCI -- analogous text could be helpful in eliminating confusion in VCI (or maybe it is there and not displaying...

There is currently a similar paragraph of note text in the VCI (see the screen shot above). Although it may not be exactly the same text you see in GCI, it is there when the approval form entry panel shows.

**Comment by jimmyzhen** [comment]
@selinad,

> However, if I click to save it as Approved (orange button), I get this, which seems correct, but we may need to make certain curators aren't thinking they are saving the pathogenic....I'm not sure what to do for this, or whether we should worry right now?...Could we have some "simple feedback" on the save evaluation part saying not yet saved? When they click button, it would say saved?

There is also a paragraph of note text in the VCI being shown when the provisional form entry panel is displayed (upon saving the evaluation). You may choose to update it if it suits the needs. For anything more complicated or needing a more well thought-out user experience, I would suggest we look into that in subsequent releases if this is not a showstopper.

**Comment by selinad** [comment]
Testing steps
- add variant 21135
- add some evaluations for benign
- go to Summary
- Save summary.  @jimmyzhen this looks great -- were we going to remove "In Progress" though?  I thought it disappeared upon first save.
![image](https://user-images.githubusercontent.com/11320314/38598154-1ddd213e-3d10-11e8-840d-9f3556b5f410.png)

- go back and add another evaluation
- return to Summary and save a new one -- all looks good
- save this as Provisional -- all looks good
![image](https://user-images.githubusercontent.com/11320314/38598301-b0317f8a-3d10-11e8-8a98-7e1c371ac505.png)
- return to interpretation and make another evaluation
- return to Summary -- all looks good
- modify to LP and give reason, save again; expect to see old provisional and ability to approve it, but see this (should still let me Approve original one, right?  But, saved behavior good since there is no disease):
![image](https://user-images.githubusercontent.com/11320314/38598409-142751d6-3d11-11e8-94fe-3e7e49c89e8c.png)

- go back out and add disease
- go back to Summary -- looks good except that still can't approve 1st provisional even though I haven't saved a second Provisional
![image](https://user-images.githubusercontent.com/11320314/38598526-8bde5b34-3d11-11e8-8d59-9d3d080b49b3.png)
- save a new Provisional --- looks great
- save a first approved -- looks great!
- went back to delete Disease -- great note on modal
- changed to VUS in summary and saved.  All looked good
- saved Provisional and got correct status!
- saved as Approved

All in all looks great!!! Just the one bug when I saved as LP (see above) and there was missing ability to save the existing Prov as Approved....

Dashboard -- was I supposed to see this as Approved?  It shows none
![image](https://user-images.githubusercontent.com/11320314/38598696-4e8546b6-3d12-11e8-8ac7-9361c47577cb.png)

here is is when I go into it:
![image](https://user-images.githubusercontent.com/11320314/38598711-5caa0ac4-3d12-11e8-9b4d-5174a62809b4.png)

**Comment by wrightmw** [comment]
Previous error I reported whereby "Approved/New Provisional" was being incorrectly reported as just "Provisional" in the "All Existing Interpretations" section has now been fixed. Thanks Jimmy!
@selinad I have added some Affiliations for you in this new testing instance... you should see them next time you logout and back in again.

**Comment by wrightmw** [comment]
Jimmy and @selinad ... there is an issue with the dashboard. If I save an "Approved, New Provisional"...when I go back to the dashboard it says just "Provisional":

<img width="1164" alt="screen shot 2018-04-11 at 9 14 45 am" src="https://user-images.githubusercontent.com/15131169/38619034-3cad3da2-3d69-11e8-99a8-58c0a86931a7.png">

<img width="1159" alt="screen shot 2018-04-11 at 9 15 11 am" src="https://user-images.githubusercontent.com/15131169/38619042-44d80142-3d69-11e8-85aa-3797b1e6f25f.png">

**Comment by jimmyzhen** [comment]
@selinad,

Please see my responses to your feedback from your testing last night:

> were we going to remove "In Progress" though? I thought it disappeared upon first save.

"In progress" remains until the 1st Provisional is saved.

> modify to LP and give reason, save again; expect to see old provisional and ability to approve it, but see this (should still let me Approve original one, right?
> go back to Summary -- looks good except that still can't approve 1st provisional even though I haven't saved a second Provisional

As the way it is implemented now, I am disabling the ability to approve a saved Provisional as soon as the user saves changes in the Summary. The thinking behind it is that the user should provision the newly saved changes in the Summary again prior to approving it.

**Comment by selinad** [comment]
Hi @jimmyzhen , @wrightmw discovered a bug when we were demo'ing for the curators.  The current approved moves to archived -- it should have a green flag instead of a file box
![image](https://user-images.githubusercontent.com/11320314/38694805-7b6619de-3e3f-11e8-8ee1-1b2226b220b8.png)

**Comment by selinad** [comment]
@jimmyzhen feedback from curators: 1) loved the workflow and no confusion around it! 2) adding "current" before green flag and "archived" before file would be good -- if this is not possible, could we make the icons a tad larger -- feedback was the file box looked like a trash can =:-0

**Comment by selinad** [comment]
@jimmyzhen would it be "quick" to have the gray "View Approved Summary" (or Provisional) have a blue border?  NOT NECESSARY at all (they were fine with it being gray) but if quick it may make it more obvious it's clickable.

**Comment by selinad** [comment]
@jimmyzhen one more bug, which seems to be related to a long variant name -- problems on Dashboard display
<img width="1077" alt="screen shot 2018-04-12 at 10 47 40 am" src="https://user-images.githubusercontent.com/11320314/38695230-83fffb22-3e40-11e8-8540-417cfa0cf8cc.png">

**Comment by jimmyzhen** [comment]
@selinad, @wrightmw,

Please use the following instance to review the changes and fixes in regards to your comments and reports above:
https://jz-approval-workflow.demo.clinicalgenome.org

Thanks!

**Comment by selinad** [comment]
@jimmyzhen looks GREAT!!!!  I couldn't break it!  Nicely done.  Only 2 small items:

- it doesn't seem like the tiestamp for Saved Summary updates all the time.  I saved one at 8:27 and one at 8:29 and it remained at 8:27.

- still a wrap problem -- I don't think there is a need to stress about it now, though (24229)
![image](https://user-images.githubusercontent.com/11320314/38715733-b84d4df2-3e91-11e8-9be2-07347139953e.png)

**Comment by jimmyzhen** [comment]
@selinad, @wrightmw,

I've addressed the timestamp and word wrapping issues, and have updated the instance. Please review again:
https://jz-approval-workflow.demo.clinicalgenome.org

**Comment by selinad** [comment]
@jimmyzhen thank you!

Timestamp appears to update on last edit of an evalution (not summary), but this is good if not the best.

 I still see wrapping -- not critical, IMHO, for this release -- a long title will be unusual, but I'm not certain the time involved to address
![image](https://user-images.githubusercontent.com/11320314/38762157-0ce0eb7e-3f3e-11e8-9bd8-e7a0e41d3917.png)

**Comment by jimmyzhen** [comment]
@selinad,

I suspect this is due to the missing CSS in the test instance. It is looking okay in my local instance:
![image](https://user-images.githubusercontent.com/6956310/38762454-142c9344-3f41-11e8-993b-c24c70fd6159.png)

**Comment by selinad** [comment]
@jimmyzhen looks fantastic!  Thanks for checking it on your local instance.  The release candidate will have the missing CSS, correct?

**Comment by jimmyzhen** [comment]
@selinad,

Yes, it will.

**Comment by wrightmw** [comment]
Very minor point... in the Provisional/Approval process the field for Affiliations is named "ClinGen Affiliation" but we may wish to change this in future to just "Affiliation"... we already have some Affiliations that are not "ClinGen" and in time there will no doubt be many more.

**Comment by jimmyzhen** [comment]
@wrightmw Please create a ticket for this so that we can address it in one of the subsequent releases. Thank you.

**Comment by selinad** [comment]
@jimmyzhen unfortunatley @wrightmw and I both get this message
![image](https://user-images.githubusercontent.com/11320314/38784653-f8596856-40c9-11e8-8856-9288d568bc00.png)

when clicking on "View Provisional Summary" after saving a Provisional in the VCI
![image](https://user-images.githubusercontent.com/11320314/38784651-ecb3a6ce-40c9-11e8-9506-087e5967dd4a.png)

**Comment by selinad** [comment]
@jimmyzhen one thing I now can't remember whether I noticed or is newly introduced. 

When you have this screen (just saved a summary and there IS an associated disease)
![image](https://user-images.githubusercontent.com/11320314/38784797-2356558a-40cc-11e8-95d0-d8df5aaefbb6.png)

If you hit Edit, you get this:
![image](https://user-images.githubusercontent.com/11320314/38784799-3241b72e-40cc-11e8-9bf8-2b3ddffd1c8d.png)

The good news is the ability to save as Provisional disappears, but the need a disease note appears even though one already exists.  It's' not a big deal as when I hit Save again, it gives me provisional option....I think we can fix it for next time unless it is "quick" to remove this button.

**Comment by selinad** [comment]
@jimmyzhen this is my bad for not catching earlier -- the display here is a bit confusing:

![image](https://user-images.githubusercontent.com/11320314/38785040-4518ac56-40cf-11e8-82ec-88a0b59bc689.png)

Would it be quick to make the Provisional/Approved Status and tag be on the same line as the rest of the info for the interpretation?  If not, I'll make a ticket for next time

**Comment by jimmyzhen** [comment]
@selinad @wrightmw,

I have fixed the view provisional/approval summary problem in the VCI. Please try again and let me know there are any problems. Thanks.

**Comment by selinad** [comment]
@jimmyzhen it works!!  Thank you !

- [ ] I'm not sure I'd seen it for a variant with a CV and CA ID -- a space needed between the IDs if simple:

![image](https://user-images.githubusercontent.com/11320314/38785493-a4045a84-40d5-11e8-8587-a11086481949.png)

**Comment by jimmyzhen** [comment]
@selinad,

If we are gonna make the VCI status(es) inline with the timestamp, they may wrap to the next line (e.g. the `APPROVED` is in one line and the `NEW PROVISIONAL` is in the next) when you have something like the screenshot below:
![image](https://user-images.githubusercontent.com/6956310/38785830-8dbf33d6-40d8-11e8-9803-a502c075ef46.png)

Would you still prefer to make status(es) inline with timestamp?

**Comment by selinad** [comment]
thx  @jimmyzhen -- would it be quick to slightly increase the space between interpretations for now?

**Comment by jimmyzhen** [comment]
@selinad It will be slightly quicker.

**Comment by selinad** [comment]
Implemented in R18 -- great work @jimmyzhen !!!!
